Introducing a TimeSeries.correlate() method

alurban opened the pull request:
@duncanmmacleod, this pull request makes the following changes:

* Introduce a method, `TimeSeries.correlate()`, which convolves `self` with the phase reverse of a given signal, yielding the matched-filter SNR
* The new method whitens both the input and the matched-filter if requested, using an ASD of the input followed by two calls to `TimeSeries.whiten()`
* Add a unit test and documentation example for the new method (the latter requires `LIGO.ORG` credentials)
* Update `TimeSeries.whiten()` to normalize by expected standard deviation, rather than observed
* Update the unit test for `TimeSeries.whiten()`

Note, the SNR timeseries returned by `TimeSeries.correlate()` has the same length and timestamps as `self`. Both `self` and the matched-filter argument are detrended (to give the output zero mean), and the output SNR is normalized by its expected standard deviation (to give unit variance). Users will have to remember to take the absolute value of the output if they don't care about minus signs.

This should be considered a target for gwpy 0.13.0/1.0.0.

This fixes #892.

pep8speaks commented:
Hello @alurban! Thanks for submitting the PR.

 - In the file [`examples/timeseries/snr.py`](https://github.com/alurban/gwpy/blob/6948376f3d7e76efaeaa089f78fba09bf1ddeb8f/examples/timeseries/snr.py), following                are the PEP8 issues :
> [Line 58:1](https://github.com/alurban/gwpy/blob/6948376f3d7e76efaeaa089f78fba09bf1ddeb8f/examples/timeseries/snr.py#L58): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file


 - There are no PEP8 issues in the file [`gwpy/timeseries/tests/test_timeseries.py`](https://github.com/alurban/gwpy/blob/6948376f3d7e76efaeaa089f78fba09bf1ddeb8f/gwpy/timeseries/tests/test_timeseries.py) !

 - There are no PEP8 issues in the file [`gwpy/timeseries/timeseries.py`](https://github.com/alurban/gwpy/blob/6948376f3d7e76efaeaa089f78fba09bf1ddeb8f/gwpy/timeseries/timeseries.py) !

alurban commented:
@duncanmmacleod, the example introduced in `examples/timeseries/correlate.py` produces the following plots:

* SNR timeseries of h(t) cross-correlated with the PSL channel
![whistle_snr](https://user-images.githubusercontent.com/5273800/45530071-6ca8e280-b7ae-11e8-8e67-0399fa098998.png)

* Q-scan of h(t)
![whistle_hoft](https://user-images.githubusercontent.com/5273800/45530081-76324a80-b7ae-11e8-80cc-89cdba110fae.png)

* Q-scan of PSL photodiode
![whistle_psl](https://user-images.githubusercontent.com/5273800/45530088-81857600-b7ae-11e8-9dcf-2296714ca510.png)

@tjma12, you may also be interested in this functionality for various reasons.

alurban commented:
@duncanmmacleod, I've rebased this branch against `upstream/develop` to get all the functionality introduced in gwpy-0.12.2. Note, this PR is ready for review but still marked as WIP because it depends on #883, which should be merged first.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1601,6 +1601,87 @@ def convolve(self, fir, window='hanning'):
+                                         fduration=wduration,
+                                         highpass=highpass)
+        else:
+            in_ = self.copy()
```
why is this `copy()` required?

alurban replied:
I think I was attempting to be safe by not performing any operations in-place. Would it be better to replace `in_ = self.copy().whiten(...)` with `self.whiten(...)` on line 1674, and to replace `in_ = self.copy()` with `in_ = self` on line 1678?

alurban replied:
This is done in the most recent commit, let me know if you like this change better.

alurban commented:
@duncanmmacleod, after talking to @andrew-lundgren, I've updated this PR to more gracefully normalize both the whitening filter and the matched-filter SNR based on the _expected_ standard deviation rather than the _observed_ one. The upshot is that `TimeSeries.whiten()` and `TimeSeries.correlate()` will return Guassian-distributed data with zero mean and unit variance **only if** the input is Guassian and stationary, and will show deviations above Guassian noise elsewise. I've also updated the unit tests for `TimeSeries.whiten()` and `TimeSeries.correlate` which should now be more complete.
